tests/bsim/bluetooth/host/l2cap/einprogress/dut/src/dut.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/l2cap.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/buf.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/util_macro.h>
+
+#include "testlib/conn.h"
+#include "testlib/scan.h"
+
+#include "babblekit/flags.h"
+#include "babblekit/testcase.h"
+
+/* local includes */
+#include "data.h"
+
+LOG_MODULE_REGISTER(dut, CONFIG_APP_LOG_LEVEL);
+
+struct tester_t {
+	size_t sdu_count;
+	struct bt_conn *conn;
+	struct bt_l2cap_le_chan le_chan;
+	struct k_fifo ack_todo;
+};
+
+static atomic_t acl_pool_refs_held[CONFIG_BT_BUF_ACL_RX_COUNT];
+static struct tester_t tester;
+
+BUILD_ASSERT(IS_ENABLED(CONFIG_BT_TESTING));
+BUILD_ASSERT(IS_ENABLED(CONFIG_BT_HCI_ACL_FLOW_CONTROL));
+void bt_testing_trace_event_acl_pool_destroy(struct net_buf *destroyed_buf)
+{
+	int buf_id = net_buf_id(destroyed_buf);
+
+	__ASSERT_NO_MSG(0 <= buf_id && buf_id < ARRAY_SIZE(acl_pool_refs_held));
```

Comment:
Why `__ASSERT_NO_MSG` instead of `TEST_ASSERT`?

Reply:
I'm trying to be helpful in locating the problem for anyone debugging.

The assert is an assertion in the local code, i.e. the problem is in test itself.

OTOH `TEST_ASSERT` is checking some property about the DUT. If it fails, the DUT is not following the specification encoded by the test.